[Feat] 관리자의 이슈 수정, 삭제 기능을 구현한다.

asd42270 opened the pull request:
# Pull Request 요약

- 이슈 수정, 삭제 기능 구현

## 변경 사항

- 우선 물리 삭제로 구현했습니다!
- 추후 리팩토링을 고려하여 논리 삭제 코드는 주석으로 남겨두었습니다.
- 수정 메서드 추가 했습니다!
- 포스트맨 테스트 완료했습니다~~
- 테스트 코드 작성 완료 무사 통과
## 관련 이슈

- #145 

## 개발 유형

- [ ] 버그 수정
- [ ] 새로운 기능 개발
- [ ] 리팩토링
- [ ] 테스트 코드 작성
- [ ] 문서 업데이트
- 기타 (아래에 설명 추가)

## 작업 기간

- 작업 시작일: (2024-06-03)
- 작업 종료일: (2024-06-03)

## 유의사항

- 이슈 수정은 내일 오전 or 오늘 밤 중으로 올릴게유...(올렸어유)
- 테스트 코드는 오늘 밤 or 내일 오전 중으로...

## 참고문헌

- 이 PR을 작성하며 참조한 문헌이나 가이드가 있다면 여기에 링크를 추가하세요.

asd42270 committed:
[feat] SoftDeletableBaseTimeEntity으로 상속 클래스 변경, delete 메서드 추가

asd42270 committed:
[feat] 관리자는 이슈를 삭제할 수 있다.
- ArticleIssueController: 삭제 메서드를 추가한다.
- ArticleIssueRepository: 삭제를 위해 메서드를 추가
- ArticleIssueService: 이슈와 권한 검증 로직을 포함한 삭제 비즈니스 로직을 작성한다.

asd42270 committed:
[feat] 관리자는 이슈를 삭제할 수 있다.
- 삭제 메서드 추가에 따른 조회 메서드 변경
- ArticleIssueRepository: 조회 메서드에 DeletedAtIsNull옵션을 추가한다.
- ArticleIssueCheckService: 바뀐 메서드를 반영한다.
- ManageIssueCheckServiceTest: 메서드 변경에 따른 테스트 코드를 수정한다.
- RepairIssueCheckServiceTest: 메서드 변경에 따른 테스트 코드를 수정한다.
- 테스트 모두 통과했습니다.

asd42270 committed:
[feat] 관리자는 이슈를 삭제할 수 있다.
- ArticleIssueController: articleIssueService를 활용하기 위한 코드 변경
- ArticleIssueControllerDocsTest: 의존성 주입 추가로 인한 코드 추가
- ArticleIssueService: 물리 삭제로 우선 변경

- 포스트맨 테스트 완료, 변경된 테스트 코드 이상 없음 확인

asd42270 committed:
[test] 관리자는 이슈를 삭제할 수 있다.
- IssueDeleteServiceTest: 테스트 코드를 작성한다.

asd42270 committed:
[test] 관리자는 이슈를 삭제할 수 있다.
- 이슈 삭제의 테스트 코드와 api 문서를 작성한다.

asd42270 committed:
[test] 관리자는 이슈를 수정할 수 있다.
- IssueUpdateRequestDto: 이슈 수정을 위한 requestDto를 작성한다.
- ArticleIssue: title과 content에 setter를 추가하고, setUpdateAt을 추가한다.
- ArticleIssueService: 서비스 코드를 작성한다.

asd42270 committed:
[test] 관리자는 이슈를 수정할 수 있다.
- ArticleIssueService: 메서드 이름 변경
- ArticleIssueController: 수정 컨트롤러 추가

Galmaeki commented on src/docs/asciidoc/articleissue/issuedelete-api.adoc:
```diff
@@ -0,0 +1,23 @@
+= Issue Delete REST API Docs
```
기존 이슈 페이지랑 합쳐도 좋을것같슴다

asd42270 replied:
아 맞네요!! 알겠습니다!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/articleIssue/entity/ArticleIssue.java:
```diff
@@ -15,16 +15,18 @@
 @AllArgsConstructor(access = AccessLevel.PROTECTED)
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
-public class ArticleIssue extends BaseTimeEntity {
+public class ArticleIssue extends SoftDeletableBaseTimeEntity {
```
SoftDeletableBaseTimeEntity 는 왜 추가하셨는지 궁금합니다!

asd42270 replied:
아 초기에 논리 삭제 때문에 상속 받았었습니다!! 리팩토링 때도 사용하기 위해 살려두는 게 낫지 않을까용

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/articleIssue/entity/ArticleIssue.java:
```diff
@@ -56,4 +58,10 @@ public ArticleIssue(String title, String content, RepairArticle repairArticle,
+        super.delete(currentTime);
+    }
+
+    public void setUpdatedAt(LocalDateTime currentTime) {this.updatedAt = currentTime;}
```
이렇게 임의로 설정을 안해도 쿼리 변동시 updateAt이 자동으로 갱신됩니다!

asd42270 replied:
오 넵넵!!!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/articleIssue/repository/ArticleIssueRepository.java:
```diff
@@ -9,7 +9,9 @@
-    Optional<ArticleIssue> findFirstByRepairArticleOrderByCreatedAtDesc(RepairArticle repairArticle);
+    Optional<ArticleIssue> findFirstByRepairArticleAndDeletedAtIsNullOrderByCreatedAtDesc(RepairArticle repairArticle);
+
+    Optional<ArticleIssue> findByIdAndDeletedAtIsNull(Long id);
```
현재 이슈 히스토리가 추가될지 안될지 모르는 상황이라 DeletedAtIsNull 검증은 불필요한 것 같습니다!

asd42270 replied:
아 넵 알겠습니다!!

ghrltjdtprbs commented on src/main/java/com/final_10aeat/domain/articleIssue/repository/ArticleIssueRepository.java:
```diff
@@ -9,7 +9,9 @@
 public interface ArticleIssueRepository extends JpaRepository<ArticleIssue, Long> {
 
-    Optional<ArticleIssue> findFirstByManageArticleOrderByCreatedAtDesc(ManageArticle manageArticle);
+    Optional<ArticleIssue> findFirstByManageArticleAndDeletedAtIsNullOrderByCreatedAtDesc(ManageArticle manageArticle);
```
mysql에는 AutoIncrement 기능이 있으니 id를 기준으로 desc 를 수행하면 성능이 더 좋아진다는 점 참고해주세요!

Galmaeki replied:
보충해서 설명하면 mySql은 기본적으로 ID 기준으로 인덱싱을 진행하기 때문에 인덱싱이 안된 CreateAt보다 좋은 성능이 나옵니다

asd42270 replied:
넵 감사합니다!!

ghrltjdtprbs reviewed (changes requested):
수고하셨습니다! 일단 내일 오전까지 이슈 히스토리 기능 구현여부 나온다니까 일단 pr 머지하지 마시고 여기서 수정하면 좋을 것 같습니당 😊😊

Galmaeki reviewed (approved):
고생하셨습니다!

asd42270 committed:
[fix] 관리자는 이슈를 수정할 수 있다.
- api 문서 변경

asd42270 committed:
[fix] setUpdateAt 삭제
- ArticleIssue: 수정일의 세터를 삭제한다.

asd42270 committed:
[fix] ArticleIssueRepository와 ArticleIssueService를 수정한다.
- ArticleIssueRepository: id 기준 정렬로 변경하고, deletedAtIsNull을 주석처리한다.
- ArticleIssueService, ArticleIssueCheckService: repository변경과 엔티티 테이블 수정에 다른 서비스 로직을 수정한다.

asd42270 committed:
[test] 코드 수정에 따른 테스트 코드 수정
- ManageIssueCheckServiceTest, RepairIssueCheckServiceTest

asd42270 committed:
[test] 테스트 코드를 작성한다.
- IssueUpdateServiceTest: 서비스 테스트 코드를 작성한다.
- ArticleIssueControllerDocsTest, articleissue-api.adoc: api 문서 작성과 그에 관련한 테스트 코드를 작성하낟.
